### PR TITLE
Add driver options to swift to enable MCCAS

### PR DESCRIPTION
**Pull request by rastogishubham**
To enable MCCAS, the following driver options have been added

-cas-backend: Enable MCCAS backend in swift, the option -cache-compile-job must also be used.

-cas-backend-mode=native: Set the CAS Backend mode to emit an object file after materializing it from the CAS.

-cas-backend-mode=casid: Emit a file with the CASID for the CAS that was created.

-cas-backend-mode=verify: Verify that the object file created is identical to the object file materialized from the CAS.

-cas-emit-casid-file: Emit a .casid file next to the object file when CAS Backend is enabled.

This commit reverts 0e8554bb15408da6f0619176b6f84eadf88ab983, which reverted 3c949028e8363463c5520aab1ea49c24f5103b6c

Apologies for that, the test mccas.swift should have been guarded correctly, I think this should fix it

### CODE REVIEW

**Comment by rastogishubham** [comment]
@swift-ci please test

**Comment by cachemeifyoucan** [comment]
Just curious, what is the reason for targeting `rebranch`, but not `main`?

**Comment by rastogishubham** [comment]
@cachemeifyoucan I remember just asking Adrian what branch to put it into and he said rebranch, so I just went with that, if you think it should be in main, I can change it to that too

File: test/CAS/mccas.swift
Comment (cachemeifyoucan) [inline test/CAS/mccas.swift:1]: why `CPU=arm64`? This test should work on x86 right? Using arm64 will get zero coverage on CI.

I will just only run on macOS for now and remove `-target` flag from the tests (`%target-swift-frontend` already builds for host triple)

File: lib/Frontend/CompilerInvocation.cpp
Comment (cachemeifyoucan) [inline lib/Frontend/CompilerInvocation.cpp:1598]: This should still be `EnableCaching` as I don't think MCCAS affect anything in clangImporter.

**Review by cachemeifyoucan — APPROVED** [review]
Two comments. LGTM after that.

I actually prefer land this on main unless there is good reason to avoid `main`. We can check with Adrian and Swift release manager.

**Comment by rastogishubham** [comment]
@swift-ci please test

**Comment by rastogishubham** [comment]
@etcwilde and @bnbarham please let me know if this looks okay to merge, I do not want to break rebranch on linux again

File: test/CAS/mccas.swift
Comment (cachemeifyoucan) [inline test/CAS/mccas.swift:13]: `Jllvmcas://`? I won't use that. I assume J is coming from the VBR8 encoding and it is tied to the length of the CASID, which should not be assumed to be constant (since it is configureable) and it is also unnecessary to check that.

**Review by etcwilde — APPROVED** [review]
The test failures look fine to me. With the cut of the release/5.10 branch, PR'ing against `main` should be fine if you don't want it going into 5.10.

Address @cachemeifyoucan's comments, but the failures look fine to me.

**Comment by bnbarham** [comment]
> With the cut of the release/5.10 branch, PR'ing against main should be fine if you don't want it going into 5.10.

I assumed this was dependent on stable/20230725. If not then yeah, main would be better.

**Comment by rastogishubham** [comment]
The dependent is stable/20230725, so rebranch is correct

**Comment by rastogishubham** [comment]
@swift-ci please test

**Review by cachemeifyoucan — APPROVED** [review]

**Pull request merged**
